[GlobalISel] Improve combines for extend operation by taking hint ins…

dfszabo opened the pull request:
…tructions into account

Hint instructions like G_ASSERT_ZEXT cann be viewed as a copy. Including this fact into the combiner allows the match more patterns involving such instructions.

dfszabo commented:
Greetings @arsenm @aemerson, if you have time can you review this?

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/combine-ext.mir:
```diff
@@ -6,21 +6,41 @@ body:             |
   bb.1:
   liveins: $x0
     ; CHECK-LABEL: name: test_combine_anyext_trunc
-    ; CHECK: [[COPY:%[0-9]+]]:_(s64) = COPY $x0
```
Can you pre-commit the change from CHECK to CHECK-NEXT?

arsenm replied:
Splitting it into 2 commits in the same PR is pointless, as the merge behavior is squash and submit. You can directly push the regen to main

dfszabo replied:
Ohh apologies I didn't knew this squash behavior, I am used to bitbucket. I don't have rights to push directly to main, should I create a PR with the update test?

arsenm replied:
Yes

dfszabo replied:
Okay, created one: https://github.com/llvm/llvm-project/pull/77756

arsenm reviewed (approved):
lgtm with nit

dfszabo committed:
[GlobalISel] Improve combines for extend operation by taking hint instructions into account

Hint instructions like G_ASSERT_ZEXT can be viewed as a copy. Including this fact into the combiner allows to match more patterns involving such instructions.
